File: exams/frmResultMeanAnalysis.vb
Public Class frmResultMeanAnalysis

    Private Sub frmResultMeanAnalysis_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If Not connect() Then
            Me.Close()
        Else
            lstExaminations.Items.Clear()
            grpMultiExaminations.Enabled = True
            fill_year()
            get_term()
            load_class(cboClass)
            cboTerm.SelectedItem = term
            cboYear.SelectedItem = Today.Year
            cboSelectedTerm.SelectedItem = term
        End If
    End Sub

    Private Sub fill_year()
        cboYear.Items.Clear()
        cboSelectedYear.Items.Clear()
        fill_years(cboYear)
        fill_years(cboSelectedYear)
        cboYear.SelectedItem = Today.Year
        cboSelectedYear.SelectedItem = Today.Year
    End Sub

    Private Sub btnAddExam_Click(sender As Object, e As EventArgs) Handles btnAddExam.Click
        add_exam()
    End Sub

    Private Function exists(ByVal str As String)
        Dim i As Integer
        Dim selectedTerm As String = String.Empty

        For i = 0 To lstExaminations.Items.Count - 1
            If lstExaminations.Items.Item(i).Text = str And lstExaminations.Items.Item(i).SubItems(2).Text = cboYear.SelectedItem And lstExaminations.Items.Item(i).SubItems(1).Text = cboTerm.SelectedItem Then
                MsgBox("The Exam Already Exist In The List")
                Return True
            End If
        Next
        Return False
    End Function

    Private Sub fill_exam()
        If qread("SELECT ExamName FROM examinations WHERE Term='" & cboTerm.SelectedItem & "' AND Year='" & cboYear.SelectedItem & "'") Then
            cboExamName.Items.Clear()
            cboExamName.Items.Add(None)
            While dbreader.Read
                cboExamName.Items.Add(dbreader("ExamName"))
            End While
            cboExamName.SelectedItem = None
            dbreader.Close()
        Else
            failure("Could Not Load Examinations!")
        End If
    End Sub

    Private Sub add_exam()
        If cboExamName.SelectedItem <> None Then
            If Not exists(cboExamName.SelectedItem) Then
                Dim li As New ListViewItem
                li = lstExaminations.Items.Add(cboExamName.SelectedItem)
                li.SubItems.Add(cboTerm.SelectedItem)
                li.SubItems.Add(cboYear.SelectedItem)
            End If
        End If
    End Sub

    Private Sub cboYear_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboYear.SelectedIndexChanged
        If cboTerm.SelectedItem <> None And cboYear.SelectedItem.ToString <> None Then
            fill_exam()
        End If
    End Sub

    Private Sub cboTerm_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboTerm.SelectedIndexChanged
        If cboTerm.SelectedItem <> None And cboYear.SelectedItem.ToString <> None Then
            fill_exam()
        End If
    End Sub

    Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        Me.Close()
    End Sub

    Private Sub btnEnterMarks_Click(sender As Object, e As EventArgs) Handles btnEnterMarks.Click
        If lstExaminations.Items.Count = 0 Then
            MsgBox("There Are No Exams To Analyze")
        ElseIf cboClass.SelectedItem = Nothing Then
            MsgBox("Please Select The Class")
            'todo add condition to check whether grading scheme exists
        ElseIf cboSortBy.SelectedItem = String.Empty And cboSortBy.Visible Then
            MsgBox("Please Choose The Mode To Sort The Result Analysis!")
        Else

            If rdSubjectBased.Checked = True Then
                gradingType = "SubjectBased"
            Else
                gradingType = "ClassBased"
            End If

            selectedTerm = cboSelectedTerm.SelectedItem.ToString()
            selectedClass = cboClass.SelectedItem.ToString()
            selectedYear = cboSelectedYear.SelectedItem.ToString()
            sortGradesBy = cboSortBy.SelectedItem.ToString()

            tm = cboTerm.SelectedItem
            yr = cboYear.SelectedItem

            examList = New List(Of Tuple(Of String, String, String))

            For i = 0 To lstExaminations.Items.Count - 1
                examList.Add(New Tuple(Of String, String, String)(lstExaminations.Items.Item(i).Text, lstExaminations.Items.Item(i).SubItems(1).Text, lstExaminations.Items.Item(i).SubItems(2).Text))
            Next

            subjectColumns = New List(Of String)
            qread("SHOW COLUMNS FROM subjects_done;")
            Dim columns As String = String.Empty
            While dbreader.Read
                columns = dbreader("Field")

                If columns <> "ADMNo" Then
                    subjectColumns.Add(columns)
                End If
            End While

            Me.Close()
            Dim meanResultForm As New frmMeanResults
            meanResultForm.MdiParent = frmMainForm
            meanResultForm.Show()

        End If
    End Sub


End Class

File: exams/frmResultMeanAnalysis.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace exams
{
    public partial class frmResultMeanAnalysis
    {
        public frmResultMeanAnalysis()
        {
            InitializeComponent();
            _cboYear.Name = "cboYear";
            _cboTerm.Name = "cboTerm";
            _btnAddExam.Name = "btnAddExam";
            _btnEnterMarks.Name = "btnEnterMarks";
            _btnCancel.Name = "btnCancel";
        }

        private void frmResultMeanAnalysis_Load(object sender, EventArgs e)
        {
            if (!publicSubsNFunctions.connect())
            {
                Close();
            }
            else
            {
                lstExaminations.Items.Clear();
                grpMultiExaminations.Enabled = true;
                fill_year();
                publicSubsNFunctions.get_term();
                var argcbo = cboClass;
                publicSubsNFunctions.load_class(ref argcbo);
                cboClass = argcbo;
                cboTerm.SelectedItem = publicSubsNFunctions.term;
                cboYear.SelectedItem = DateAndTime.Today.Year;
                cboSelectedTerm.SelectedItem = publicSubsNFunctions.term;
            }
        }

        private void fill_year()
        {
            cboYear.Items.Clear();
            cboSelectedYear.Items.Clear();
            var argcbo = cboYear;
            publicSubsNFunctions.fill_years(ref argcbo);
            cboYear = argcbo;
            var argcbo1 = cboSelectedYear;
            publicSubsNFunctions.fill_years(ref argcbo1);
            cboSelectedYear = argcbo1;
            cboYear.SelectedItem = DateAndTime.Today.Year;
            cboSelectedYear.SelectedItem = DateAndTime.Today.Year;
        }

        private void btnAddExam_Click(object sender, EventArgs e)
        {
            add_exam();
        }

        private object exists(string str)
        {
            int i;
            string selectedTerm = string.Empty;
            var loopTo = lstExaminations.Items.Count - 1;
            for (i = 0; i <= loopTo; i++)
            {
                if (Conversions.ToBoolean(Operators.AndObject(Operators.AndObject((lstExaminations.Items[i].Text ?? "") == (str ?? ""), Operators.ConditionalCompareObjectEqual(lstExaminations.Items[i].SubItems[2].Text, cboYear.SelectedItem, false)), Operators.ConditionalCompareObjectEqual(lstExaminations.Items[i].SubItems[1].Text, cboTerm.SelectedItem, false))))
                {
                    Interaction.MsgBox("The Exam Already Exist In The List");
                    return true;
                }
            }

            return false;
        }

        private void fill_exam()
        {
            string argq = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject("SELECT ExamName FROM examinations WHERE Term='", cboTerm.SelectedItem), "' AND Year='"), cboYear.SelectedItem), "'"));
            if (publicSubsNFunctions.qread(ref argq))
            {
                cboExamName.Items.Clear();
                cboExamName.Items.Add(publicSubsNFunctions.None);
                while (publicSubsNFunctions.dbreader.Read())
                    cboExamName.Items.Add(publicSubsNFunctions.dbreader["ExamName"]);
                cboExamName.SelectedItem = publicSubsNFunctions.None;
                publicSubsNFunctions.dbreader.Close();
            }
            else
            {
                publicSubsNFunctions.failure("Could Not Load Examinations!");
            }
        }

        private void add_exam()
        {
            if (Conversions.ToBoolean(Operators.ConditionalCompareObjectNotEqual(cboExamName.SelectedItem, publicSubsNFunctions.None, false)))
            {
                if (Conversions.ToBoolean(!exists(Conversions.ToString(cboExamName.SelectedItem))))
                {
                    var li = new ListViewItem();
                    li = lstExaminations.Items.Add(cboExamName.SelectedItem);
                    li.SubItems.Add(cboTerm.SelectedItem);
                    li.SubItems.Add(cboYear.SelectedItem);
                }
            }
        }

        private void cboYear_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false), (cboYear.SelectedItem.ToString() ?? "") != (publicSubsNFunctions.None ?? ""))))
            {
                fill_exam();
            }
        }

        private void cboTerm_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false), (cboYear.SelectedItem.ToString() ?? "") != (publicSubsNFunctions.None ?? ""))))
            {
                fill_exam();
            }
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void btnEnterMarks_Click(object sender, EventArgs e)
        {
            if (lstExaminations.Items.Count == 0)
            {
                Interaction.MsgBox("There Are No Exams To Analyze");
            }
            else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboClass.SelectedItem, null, false)))
            {
                Interaction.MsgBox("Please Select The Class");
            }
            // todo add condition to check whether grading scheme exists
            else if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, string.Empty, false), cboSortBy.Visible)))
            {
                Interaction.MsgBox("Please Choose The Mode To Sort The Result Analysis!");
            }
            else
            {
                if (rdSubjectBased.Checked == true)
                {
                    publicSubsNFunctions.gradingType = "SubjectBased";
                }
                else
                {
                    publicSubsNFunctions.gradingType = "ClassBased";
                }

                publicSubsNFunctions.selectedTerm = cboSelectedTerm.SelectedItem.ToString();
                publicSubsNFunctions.selectedClass = cboClass.SelectedItem.ToString();
                publicSubsNFunctions.selectedYear = cboSelectedYear.SelectedItem.ToString();
                publicSubsNFunctions.sortGradesBy = cboSortBy.SelectedItem.ToString();
                publicSubsNFunctions.tm = Conversions.ToString(cboTerm.SelectedItem);
                publicSubsNFunctions.yr = Conversions.ToInteger(cboYear.SelectedItem);
                publicSubsNFunctions.examList = new List<Tuple<string, string, string>>();
                for (int i = 0, loopTo = lstExaminations.Items.Count - 1; i <= loopTo; i++)
                    publicSubsNFunctions.examList.Add(new Tuple<string, string, string>(lstExaminations.Items[i].Text, lstExaminations.Items[i].SubItems[1].Text, lstExaminations.Items[i].SubItems[2].Text));
                publicSubsNFunctions.subjectColumns = new List<string>();
                string argq = "SHOW COLUMNS FROM subjects_done;";
                publicSubsNFunctions.qread(ref argq);
                string columns = string.Empty;
                while (publicSubsNFunctions.dbreader.Read())
                {
                    columns = Conversions.ToString(publicSubsNFunctions.dbreader["Field"]);
                    if (columns != "ADMNo")
                    {
                        publicSubsNFunctions.subjectColumns.Add(columns);
                    }
                }

                Close();
                var meanResultForm = new frmMeanResults();
                meanResultForm.MdiParent = My.MyProject.Forms.frmMainForm;
                meanResultForm.Show();
            }
        }
    }
}